KEYCLOAK-4858: Slow query performance for client with large data volume

glavoie opened the pull request:
- Changing RESOURCE_SERVER PK to the client ID.
- Changing FK on children of RESOURCE_SERVER.
- Use direct fetch of ResourceServer through ID/PK to avoid a lot of implicit Hibernate flush.

As discussed https://github.com/keycloak/keycloak/pull/4095, this is a new version of the code to improve the performance of fetching the ResourceServer through the client ID. Large number of files have been touched but it's mostly to replace the calls to get/setClientId to get/setId and change the PK on RESOURCE_SERVER to the client ID. 

Includes Liquibase changeset to migrate the keys in all children tables + re-wire the indices/unique keys/FKs.

Commit:
KEYCLOAK-4858: Slow query performance for client with large data volume
- Changing RESOURCE_SERVER PK to the client ID.
- Changing FK on children of RESOURCE_SERVER.
- Use direct fetch of ResourceServer through ID/PK to avoid a lot of implicit Hibernate flush.

glavoie commented:
I will look at this on Tuesday, when I'm back to work.

pedroigor commented:
@glavoie, which database you have used to test these changes ?

I'm having issues migrating using MySQL.

pedroigor committed:
[KEYCLOAK-4858] - Migration configuration for resource server pk changes

pedroigor commented:
@glavoie, could you please take a look at some changes I did to your PR [1] ?

If you could merge my changes I can re-test this PR and prepare to merge it.

[1] https://github.com/pedroigor/keycloak/tree/KEYCLOAK-4858

hmlnarik commented on model/jpa/src/main/resources/META-INF/jpa-changelog-3.4.0.xml:
```diff
@@ -0,0 +1,88 @@
+  ~ limitations under the License.
+  -->
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+    <changeSet author="keycloak" id="3.4.0-res-server-pk">
```
Please set author correctly

glavoie replied:
@hmlnarik, I can change that, but I looked at earlier changesets (especially those submitted by me earlier) and the author was changed to "keycloak". Haven't looked why/when. That's why I set it to "keycloak".

hmlnarik commented on model/jpa/src/main/resources/META-INF/jpa-changelog-3.4.0.xml:
```diff
@@ -0,0 +1,88 @@
+  ~ limitations under the License.
+  -->
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+    <changeSet author="keycloak" id="3.4.0-res-server-pk">
```
Please set author correctly

pedroigor replied:
I've fixed this one in my branch ...

hmlnarik commented on model/jpa/src/main/resources/META-INF/jpa-changelog-3.4.0.xml:
```diff
@@ -0,0 +1,88 @@
+        </addColumn>
+
+        <sql>
+            UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID);
```
Syntax of the statements has to work for all of
* MySQL
* Postgres
* MSSQL
* DB2
* Oracle
Please verify that it really does. This needs to be fixed before any merging can take place.

hmlnarik reviewed (changes requested):
Hi, thanks for the changes. I took a quick glance at it and overall like it but it needs very careful inspection in <sql> part of the changeset, see inline.

glavoie commented:
@pedroigor, I installed MySQL earlier and found out the same things you've fixed in your changeset. I'm pulling it and updating the PR right away. Thanks for your help. :)

pedroigor commented:
@hmlnarik @glavoie, other databases look good so far. Will perform some more tests friday/next monday.

hmlnarik commented:
@pdrozd Could you please check this PR for supported database compatibility?

vramik commented on model/jpa/src/main/resources/META-INF/jpa-changelog-authz-3.4.0.CR1.xml:
```diff
@@ -0,0 +1,93 @@
+        </addColumn>
+
+        <sql>
+            UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID);
```
This SQL doesn't work on MSSQL:
ERROR [org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider] (ServerService Thread Pool -- 55) Change Set META-INF/jpa-changelog-authz-3.4.0.CR1.xml::authz-3.4.0.CR1-resource-server-pk-change::psilva@redhat.com failed.  Error: Incorrect syntax near 'p'. [Failed SQL: UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID)]: liquibase.exception.DatabaseException: Incorrect syntax near 'p'. [Failed SQL: UPDATE RESOURCE_SERVER_POLICY p SET RESOURCE_SERVER_CLIENT_ID=(SELECT CLIENT_ID FROM RESOURCE_SERVER s WHERE s.ID = p.RESOURCE_SERVER_ID)]

glavoie replied:
I will check the syntax, I have a MSSQL instance readily available.

hmlnarik replied:
You might need to split the current changeset into three, extracting the <sql> part one into a separate one, and making variants of that changeset via <precondition> on <dbms> similarly to what is already used for db2 in https://github.com/hmlnarik/keycloak/blob/master/model/jpa/src/main/resources/META-INF/jpa-changelog-1.8.0.xml#L21-L25

glavoie replied:
This is what I fear. I'm used to create Liquibase changeset twice to support both MSSQL and Oracle.

glavoie replied:
@hmlnarik, if I am to create an exception for a specific DBMS, do you want a separate changelog file for that dbms or two separate changesets in the same file?

hmlnarik replied:
@glavoie Thanks, two separate changesets in the same file would do.

glavoie replied:
@hmlnarik, will do this then. I've confirmed that 2 changesets will be required. My DBA helped me with tests on multiple DBMS and we can't find a portable solution. For now the exception would be "Oracle", but I would need your help to see what DB2 accepts. Will likely update the PR next Monday.

hmlnarik replied:
@glavoie Certainly, please just let me know when you consider the PR ready

glavoie replied:
@hmlnarik, just updated the PR again.

Commit:
KEYCLOAK-4858: ResourceServer PK change to CLIENT_ID.
- MSSQL needs the index to be dropped before the column.
- Different UPDATE statement format to support MSSQL.

glavoie commented on model/jpa/src/main/resources/META-INF/jpa-changelog-authz-3.4.0.CR1.xml:
```diff
@@ -17,7 +17,7 @@
 
 <databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.2.xsd">
-    <changeSet author="psilva@redhat.com" id="authz-3.4.0.CR1-resource-server-pk-change">
+    <changeSet author="glavoie@gmail.com" id="authz-3.4.0.CR1-resource-server-pk-change-part1">
```
@pedroigor, put back my e-mail address here if you don't mind. ;)

glavoie commented on model/jpa/src/main/resources/META-INF/jpa-changelog-authz-3.4.0.CR1.xml:
```diff
@@ -28,13 +28,32 @@
+    </changeSet>
+    <changeSet author="glavoie@gmail.com" id="authz-3.4.0.CR1-resource-server-pk-change-part2-mssql">
+        <preConditions onSqlOutput="TEST" onFail="MARK_RAN">
+            <dbms type="mssql" />
```
@hmlnarik, after some testing, my DBA told me the Oracle syntax works on both PostgreSQL and MySQL, but the MSSQL syntax only on PostgreSQL. I did the exception for MSSQL. DB2 remains.

vramik replied:
There is DB2 result: 
ERROR [org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider] (ServerService Thread Pool -- 50) Change Set META-INF/jpa-changelog-authz-3.4.0.CR1.xml::authz-3.4.0.CR1-resource-server-pk-change-part3::glavoie@gmail.com failed.  Error: DB2 SQL Error: SQLCODE=-478, SQLSTATE=42893, SQLERRMC=COLUMN;PK_RESOURCE_SERVER;PRIMARY KEY, DRIVER=4.19.26 [Failed SQL: ALTER TABLE DBALLO14.RESOURCE_SERVER RENAME COLUMN CLIENT_ID TO ID]: liquibase.exception.DatabaseException: DB2 SQL Error: SQLCODE=-478, SQLSTATE=42893, SQLERRMC=COLUMN;PK_RESOURCE_SERVER;PRIMARY KEY, DRIVER=4.19.26 [Failed SQL: ALTER TABLE DBALLO14.RESOURCE_SERVER RENAME COLUMN CLIENT_ID TO ID]

glavoie replied:
@vramik, pushed a new version that drops all FK/PK/indices, renames the columns, then adds everything back. Still works on MySQL/MSSQL/Oracle. Any better on DB2?

vramik commented:
Thnak you @glavoie, DB2 seems to work now. 
@hmlnarik from POV the PR is ready to merge, all DBs passed with latest changes.

glavoie commented:
Thanks for the merge!
